Add references for Chai Matchers and Network Helpers (#2763)
NomicFoundation/hardhat: docs/network-helpers/reference.md

# Reference

## Mining blocks

### `mine([blocks], [options])`

Mines a specified number of blocks at a given interval.

Parameters:

- `blocks`: Number of blocks to mine. Defaults to 1.
- `options.interval`: Configures the interval (in seconds) between the timestamps of each mined block. Defaults to 1.

Example:

```ts
// mine a new block
await helpers.mine();

// mine several blocks
await helpers.mine(1000);

// mine several blocks with a given interval between their timestamps
await helpers.mine(1000, { interval: 15 });
```

### `mineUpTo(blockNumber)`

Mines new blocks until the latest block number is `blockNumber`.

Parameters:

- `blockNumber`: Must be greater than the latest block's number.

Example:

```ts
await helpers.mineUpTo(1234);
```

## Manipulating accounts

### `setBalance(address, balance)`

Sets the balance for the given address.

Parameters:

- `address`: The address whose balance will be edited.
- `balance`: The new balance to set for the given address, in wei.

Example:

```ts

> await helpers.setBalance(address, 100n**18n);

Author: Hmm, maybe, but I'm not sure how to do this without making the jsdoc too loaded.

```

### `setCode(address, code)`

> Modifies the bytecode stored at an account's address.

Author: 420fc1e0589f809dd10350a91999432fe769aea0

- `address`: The address where the given code should be stored.
- `code`: The code to store.

Example:

```ts
await helpers.setCode(address, "0x1234...");
```

### `setNonce(address, nonce)`

Modifies an account's nonce by overwriting it.

Parameters:

- `address`: The address whose nonce is to be changed.
- `nonce`: The new nonce.

Example:

```ts
await helpers.setNonce(address, 100);
```

### `setStorageAt(address, index, value)`

Writes a single position of an account's storage.

Parameters:

- `address`: The address where the code should be stored.
- `index`: The index in storage.
- `value`: The value to store.

Example:

```ts
await helpers.setStorageAt(address, storageSlot, newValue);
```

### `getStorageAt(address, index, [block])`

Retrieves the data located at the given address, index, and block number.

Parameters:

- `address`: The address to retrieve storage from.
- `index`: The position in storage.
- `block`: The block number, or one of `"latest"`, `"earliest"`, or `"pending"`. Defaults to `"latest"`.

Returns: A string containing the hexadecimal value retrieved.

Example:

```ts
await helpers.getStorageAt(address, storageSlot);
```

### `impersonateAccount(address)`

> Allows Hardhat Network to sign transactions as the given address.

Author: 420fc1e0589f809dd10350a91999432fe769aea0

```

### `stopImpersonatingAccount(address)`

Stops Hardhat Network from impersonating the given address.

Parameters:

- `address`: The address to stop impersonating.

Example:

```ts
await helpers.stopImpersonatingAccount(address);
```

## Time helpers

### `latest()`

Returns the timestamp of the latest block.

Example:

```ts
await helpers.time.latest();
```

### `latestBlock()`

Returns the number of the latest block.

Example:

```ts
await helpers.time.latestBlock();
```

### `increase(amountInSeconds)`

Mines a new block whose timestamp is `amountInSeconds` after the latest block's timestamp.

Parameters:

- `amountInSeconds`: Number of seconds to increase the next block's timestamp by.

Returns: The timestamp of the mined block.

Example:

```ts
// advance time by one hour and mine a new block
await helpers.time.increase(3600);
```

### `increaseTo(timestamp)`

Mines a new block whose timestamp is `timestamp`.

Parameters:

- `timestamp`: Must be bigger than the latest block's timestamp.

Example:

```ts
await helpers.time.increaseTo(newTimestamp);
```

### `setNextBlockTimestamp(timestamp)`

Sets the timestamp of the next block but doesn't mine one.

Parameters:

> - `timestamp`: Can be `Date` or Epoch seconds. Must be greater than the latest block's timestamp.

Reviewer: Might be good to add a link to some authoritative source on what "Epoch seconds" are.

Also, do the other helpers (which accept timestamps) also all accept `Date`? If so, would be good to mention it in the context of those other helpers.

Author: Added a link to wikipedia, and opened [an issue in Linear](https://linear.app/nomic-foundation/issue/HH-731/accept-date-instances-in-all-the-network-helpers-that-accept) for the `Date` thing.

Example:

```ts
// set the timestamp of the next block but don't mine a new block
await helpers.time.setNextBlockTimestamp(newTimestamp);
```

## Snapshots

### `takeSnapshot()`

Takes a snapshot of the state of the blockchain at the current block.

> Returns: An object with a `restore` method that can be used to reset the network to this state.

Reviewer: ```suggestion
Returns: An object with a `restore` method that can be used to reset the network to the state in the snapshot.
```

Author: 420fc1e0589f809dd10350a91999432fe769aea0

Example:

```ts
// take a snapshot of the current state of the blockchain
const snapshot = await helpers.takeSnapshot();

// after doing some changes, you can restore to the state of the snapshot
await snapshot.restore();
```

## Fixtures

### `loadFixture()`

> Executes the given function and takes a snapshot of the blockchain. Each time `loadFixture` is called again with the same function, the blockchain will be restored to that snapshot instead of executing the function again.

Reviewer: ```suggestion
Executes the given function and takes a snapshot of the blockchain. Upon subsequent calls to `loadFixture` with the same function, rather than executing the function again, the blockchain will be restored to that snapshot.
```

Author: 420fc1e0589f809dd10350a91999432fe769aea0

> Useful for `beforeEach` hooks that setup the desired state of the network.

Reviewer: I would consider moving this paragraph to being the first one in this section. I was kinda lost, and felt out of context, when I began reading this section.

Author: 420fc1e0589f809dd10350a91999432fe769aea0
